src/features/soft_navigations/aggregate/interaction.js

```diff
@@ -0,0 +1,159 @@
+import { getInfo } from '../../../common/config/config'
+import { globalScope, initialLocation } from '../../../common/constants/runtime'
+import { generateUuid } from '../../../common/ids/unique-id'
+import { addCustomAttributes, getAddStringContext, nullable, numeric } from '../../../common/serialize/bel-serializer'
+import { cleanURL } from '../../../common/url/clean-url'
+import { NODE_TYPE, INTERACTION_STATUS, INTERACTION_TYPE, API_TRIGGER_NAME } from '../constants'
+import { BelNode } from './bel-node'
+
+/**
+ * link https://github.com/newrelic/nr-querypack/blob/main/schemas/bel/7.qpschema
+ **/
+export class Interaction extends BelNode {
+  id = generateUuid() // unique id that is serialized and used to link interactions with errors
+  initialPageURL = initialLocation
+  oldURL = '' + globalScope?.location
+  newURL = '' + globalScope?.location
+  customName
+  customAttributes = {}
+  customDataByApi = {}
+  queueTime // only used by initialPageLoad interactions
+  appTime // only used by initialPageLoad interactions
+  newRoute
+  /** Internal state of this interaction: in-progress, finished, or cancelled. */
+  status = INTERACTION_STATUS.IP
+  domTimestamp = 0
+  historyTimestamp = 0
+  createdByApi = false
+  keepOpenUntilEndApi = false
+  onDone = []
+  cancellationTimer
+
+  constructor (agentIdentifier, uiEvent, uiEventTimestamp, currentRouteKnown) {
+    super(agentIdentifier)
+    this.belType = NODE_TYPE.INTERACTION
+    this.trigger = uiEvent
+    this.start = uiEventTimestamp
+    this.oldRoute = currentRouteKnown
+    this.eventSubscription = new Map([
+      ['finished', []],
+      ['cancelled', []]
+    ])
+    this.forceSave = this.forceIgnore = false
+    if (this.trigger === API_TRIGGER_NAME) this.createdByApi = true
+  }
+
+  updateDom (timestamp) {
+    this.domTimestamp = (timestamp || performance.now()) // default timestamp should be precise for accurate isActiveDuring calculations
+  }
+
+  updateHistory (timestamp, newUrl) {
+    this.newURL = newUrl || '' + globalScope?.location
+    this.historyTimestamp = (timestamp || performance.now())
+  }
+
+  seenHistoryAndDomChange () {
+    return this.historyTimestamp > 0 && this.domTimestamp > this.historyTimestamp // URL must change before DOM does
+  }
+
+  on (event, cb) {
+    if (!this.eventSubscription.has(event)) throw new Error('Cannot subscribe to non pre-defined events.')
+    if (typeof cb !== 'function') throw new Error('Must supply function as callback.')
```

Comment:
Just wanna double check that we want to throw errors here instead of warns()

Reply:
I think so, yeah. This wouldn't be a user related error so warning them only confuses them. `on` is an internal fn. It'd be a dev error, and we can detect this much faster/earlier this way. Moral of the story: don't do `interaction.on('whateveryouwant', ()=>{...})` without adding 'whateveryouwant' to the `eventSubscription` map.